quota/quota.go

```diff
@@ -0,0 +1,315 @@
+	// empty then the underlying slice may not exist on the system.
+	Snaps []string `json:"snaps,omitempty"`
+}
+
+// MemoryExhaustionBehavior is the behavior determining what the system should
+// do when memory available to a quota group is exhausted.
+type MemoryExhaustionBehavior int
+
+const (
+	// DefaultOOMKiller is the default behavior to invoke to choose which
+	// processes when the memory quota resource limit is reached, and it
+	// consists of letting the Linux kernel's oom-killer run and decide, which
+	// is available on all cgroups version systems, but is not deterministic as
+	// to which processes are killed.
+	DefaultOOMKiller MemoryExhaustionBehavior = iota
+)
+
+// SliceFileName returns the name of the slice file that should be used for this
+// quota group. This name will include all of the group's parents in the name.
+// For example, a group named "bar" that is a child of the "foo" group will have
+// a systemd slice name as "foo-bar.slice". Note that the slice name may differ
+// from the snapd friendly group name, mainly in the case that the group is a
+// sub group.
+func (grp *Group) SliceFileName() string {
+	if grp.ParentGroup == "" {
+		// root group name, then the slice unit is just "<name>.slice"
+		return fmt.Sprintf("%s.slice", grp.Name)
+	}
+
+	// otherwise we need to track back to get all of the parent elements
+	grpNames := []string{}
+	parentGrp := grp.parentGroup
+	for parentGrp != nil {
+		grpNames = append([]string{parentGrp.Name}, grpNames...)
+		parentGrp = parentGrp.parentGroup
+	}
+
+	buf := &bytes.Buffer{}
+	for _, parentGrpName := range grpNames {
+		fmt.Fprintf(buf, "%s-", parentGrpName)
+	}
+	fmt.Fprintf(buf, "%s.slice", grp.Name)
+	return buf.String()
+}
+
+// allow only alphanumerics at least one character long, and not starting with
+// a number
+var validGroupNameRegexp = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9]*$`)
+
+func (grp *Group) validate() error {
+	// check that the name is a simple alphanumeric name
+	if !validGroupNameRegexp.MatchString(grp.Name) {
+		return fmt.Errorf("group name %q contains invalid characters (valid names are alphanumeric starting with a letter)", grp.Name)
+	}
+
+	// check if the name is reserved for future usage
+	switch grp.Name {
+	case "root", "system", "snapd", "user":
+		return fmt.Errorf("group name %q reserved", grp.Name)
+	}
+
+	if grp.MemoryLimit == 0 {
+		return fmt.Errorf("group memory limit must be non-zero")
+	}
+
+	// TODO: probably there is a minimum amount of bytes here that is
+	// technically usable/enforcable, should we check that too?
+
+	// check that if this is a sub-group, then the parent group has enough space
+	// to accommodate this new group (we assume that other existing sub-groups
+	// in the parent group have already been validated)
+	if grp.parentGroup != nil {
+		totalAvailable := quantity.Size(0)
+		for _, child := range grp.parentGroup.subGroups {
+			totalAvailable += child.MemoryLimit
```

Comment:
Nitpick, I think totalAvailable is slightly confusing, maybe `totalChildMemory`? IMHO this would read better in the line below:
`if totalChildMemory +grp.MemoryLimit > grp.parentGroup.MemoryLimit {...`

Reply:
I ended up going with Maciej's suggestion from the other PR, hopefully this makes sense